lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,582 @@
+		AU.setPreservesAll();
+	}
+
+	void PredicateAbstraction::printInvars (Module &M, HornDbModel &origModel)
+	{
+		 for (auto &F : M) printInvars (F, origModel);
+	}
+
+	void PredicateAbstraction::printInvars(Function &F, HornDbModel &origModel)
+	{
+		  if (F.isDeclaration ()) return;
+
+		  HornifyModule &hm = getAnalysis<HornifyModule> ();
+		  outs () << "Function: " << F.getName () << "\n";
+
+		  // -- not used for now
+		  Expr summary = hm.summaryPredicate (F);
+
+		  ZFixedPoint<EZ3> fp = *m_fp;
+
+		  for (auto &BB : F)
+		  {
+			if (!hm.hasBbPredicate (BB)) continue;
+
+			Expr bbPred = hm.bbPredicate (BB);
+
+			outs () << *bind::fname (bbPred) << ":";
+			const ExprVector &live = hm.live (BB);
+			//Expr invars = fp.getCoverDelta (bind::fapp (bbPred, live));
+			Expr invars = origModel.getDef(bind::fapp(bbPred, live));
+
+			if (isOpX<AND> (invars))
+			{
+			  outs () << "\n\t";
+			  for (size_t i = 0; i < invars->arity (); ++i)
+				outs () << "\t" << *invars->arg (i) << "\n";
+			}
+			else
+			  outs () << " " << *invars << "\n";
+		  }
+	}
+
+	HornClauseDB PredicateAbstractionAnalysis::generateAbstractDB(HornClauseDB &db, PredAbsHornModelConverter &converter)
+	{
+		HornClauseDB new_DB(db.getExprFactory ());
+
+		generateAbstractRelations(db, new_DB, converter);
+
+		generateAbstractRules(db, new_DB, converter);
+
+		generateAbstractQueries(db, new_DB);
+
+		LOG("pabs-debug", outs() << "NEW DB: \n";);
+		LOG("pabs-debug", outs() << new_DB << "\n";);
+
+		converter.setAbsDB(new_DB); //set converter
+
+		return new_DB;
+	}
+
+	void PredicateAbstractionAnalysis::generateAbstractRelations(HornClauseDB &db, HornClauseDB &new_DB, PredAbsHornModelConverter &converter)
+	{
+		//For each relation, generate its abstract version
+		for(Expr rel : db.getRelations())
+		{
+			LOG("pabs-debug", outs() << "OLD REL: " << *rel << "\n";);
+			ExprVector new_args;
+			//Push fdecl name
+			Expr old_fdecl_name = bind::fname(rel);
+			std::ostringstream oss;
+			oss << old_fdecl_name << "_pabs";
+			new_args.push_back(old_fdecl_name);
+			//Push boolean types
+			ExprVector term_vec = m_currentCandidates.find(rel)->second;
+			if(term_vec.size() > 1 || term_vec.size() == 1 && !isOpX<TRUE>(term_vec[0]))
+			{
+				for(int i=0; i<term_vec.size(); i++)
+				{
+					new_args.push_back(mk<BOOL_TY>(rel->efac()));
+				}
+			}
+			else // the candidate term is just 'true' or 'false'
+			{
+				for (int i=1; i<=bind::domainSz(rel); i++ )
+				{
+					new_args.push_back(bind::domainTy(rel, i));
+				}
+			}
+			//Push return type
+			new_args.push_back(bind::rangeTy(rel));
+			Expr new_rel = mknary<FDECL>(new_args);
+
+			//new pred name
+			Expr new_fdecl_name = mkTerm<std::string>(oss.str(), new_rel->efac());
+			new_rel = bind::rename(new_rel, new_fdecl_name);
+
+			LOG("pabs-debug", outs() << "NEW REL: " << *new_rel << "\n";);
+			new_DB.registerRelation(new_rel);
+
+			m_oldToNewPredMap.insert(std::pair<Expr, Expr>(rel, new_rel));
```

Comment:
agreed.

you can also tag a decl instead of a name. Note that name can be an
arbitrary expression. There are no restrictions.

cheers
arie
On Wed, Aug 24, 2016 at 20:18 Chenguang-Zhu <notifications@github.com>
wrote:

> In lib/seahorn/PredicateAbstraction.cc
> <https://github.com/seahorn/seahorn/pull/43#discussion_r76162815>:
>
> > +				{
> > +					new_args.push_back(bind::domainTy(rel, i));
> > +				}
> > +			}
> > +			//Push return type
> > +			new_args.push_back(bind::rangeTy(rel));
> > +			Expr new_rel = mknary<FDECL>(new_args);
> > +
> > +			//new pred name
> > +			Expr new_fdecl_name = mkTerm<std::string>(oss.str(), new_rel->efac());
> > +			new_rel = bind::rename(new_rel, new_fdecl_name);
> > +
> > +			LOG("pabs-debug", outs() << "NEW REL: " << *new_rel << "\n";);
> > +			new_DB.registerRelation(new_rel);
> > +
> > +			m_oldToNewPredMap.insert(std::pair<Expr, Expr>(rel, new_rel));
>
> We can use tag variables to create predicates names like P!pabs. But in
> this situation the map seems still useful. Because when given an abstract
> predicate P!pabs, although we can use tag expression to easily get the name
> of concrete predicate P, we still need to know the arguments list to
> construct the concrete predicate. Without this map, we must go over all
> predicates in DB and find the one that has the name P. I think maybe it's
> reasonable to keep these maps?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/seahorn/seahorn/pull/43/files/e2ed48c0dfa705fd7cb3be22af15251bdb152e84#r76162815>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFzN52SEbKl383IXWz3hjAw41Cbxm3Gmks5qjN9MgaJpZM4JsaQ0>
> .
>